### ISSUE TITLE
Refactoring support?

### ISSUE THREAD

**Issue by JasonGross** [ProofGeneral/PG#148]
What would it take to get project-wide refactoring support from ProofGeneral?  For example, if I remove an argument from a definition, I'd like PG to be able to remove it from all occurrences of that name.

**Comment by cpitclaudel** [comment]
A plug-in from @tlringer? :)

**Comment by hendriktews** [comment]
I don't think it would be wise to program a refactoring tool with
regular expressions in emacs lisp. Instead I would suggest to
build a parsing/pretty-printing library for Coq and Coq abstract
syntax trees. The parser is already in the Coq sources and big
parts of the pretty printing too. Once we have such a library you
can build a refactoring library on top of it that be driven from
emacs.

**Comment by Matafou** [comment]
I agree that it would not be wise to rely on regexps. We can imagine
something that replays the proofs and try some naive syntactic substitution
when coq raises an error, but It would miss in a lot of cases I guess.

P.

2017-01-20 9:32 GMT+01:00 hendriktews <notifications@github.com>:

**Comment by vzaliva** [comment]
I assume we are talking about local refactoring. Other refactoring systems like in Eclipse, for example, allow to rename a method or a class across multiple source files. I suspect PG lacks this high-level and it would be problematic to implement such refactoring.

**Comment by erikmd** [comment]
Indeed the implementation of such a refactoring in PG (either local to a file, or global to a Coq project across multiple files) would be quite challenging.
I agree with your remark that regexps would not be a good solution for that.
However, do you think that the data stored in the ".glob" files that are generated along with the ".vo" files could help devising such a feature? (IINM these ".glob" files are already used by coqdoc for producing a browsable documentation with proper hyperlinks between theories...)
Otherwise a feasible approach could indeed be the one suggested by @hendriktews above.
Best regards

**Comment by JasonGross** [comment]
Local to a single file is not terribly useful for me.  I want it to be global to all files in the project.  Indeed, it should be straightforward to determine most of the locations of the identifier by using the .glob file; the tricky part is accounting for implicit status and `@` and partially applied functions and partially applied identifiers used to define aliases with implicit arguments which may then be passed in a variety of ways.  (Higher order uses of the identifier are also tricky.  But just handling the explicit first-order uses would be great.)

**Comment by hendriktews** [comment]
> However, do you think that the data stored in the ".glob" files that

This is an interesting idea. Does anybody know what the glob
files contain or where I can find documentation about it?

Hendrik

**Comment by hendriktews** [comment]
Vadim Zaliva <notifications@github.com> writes:

> systems like in Eclipse, for example, allow to rename a method or a

BTW, does anybody know if Eclipse handles C function pointers
correctly when I change the argument order of a function?

Hendrik

**Comment by hendriktews** [comment]
@ejgallego Could SerAPI provide the necessary information for refactoring?

**Comment by vzaliva** [comment]
@hendriktews I have not seen .glob file documented. I quickly looked at some of them and they seem pretty simple text files listing character offsets of definitions and possibly references. I think it should be pretty simple to reverse-engineer full format.

Also check out section 13.5 at this link: http://flint.cs.yale.edu/cs428/coq/doc/Reference-Manual015.html

It explains how some sort of XML dump which might also be helpful.

**Comment by JasonGross** [comment]
@hendriktews I've [parsed .glob files for my bug minimizer](https://github.com/JasonGross/coq-tools/blob/3af130e3927ab5c365ffc504243fac449ee56510/import_util.py#L173).
> Does anybody know what the glob files contain

There are a few different kinds of lines in glob files.  One is of the form:
```
^R([0-9]+):([0-9]+) ([^ ]+) ([^ ]+) ([^ ]+) ([^ ]+)$
```
The first two groups define a character range (inclusive on both sides) for the reference (`R`).  The third group is the absolute library path of the file it's in, e.g., `Coq.Init.Logic`.  The fourth group is the module path within the file name, if the identifier is in a module, otherwise `<>`.  The fifth group is either the unqualified name of the reference (e.g., `false`), or a way of describing the notation (e.g., `:type_scope:x_'->'_x`).  The sixth group is the type of reference (`lib`rary, `ind`uctive, `var`iable, `constr`uctor, `not`ation, `syndef` (syntactic definition / alias), `def`inition)*.

There are also a few lines that do not begin with `R`.  These are:
- the first line, which is of the form `DIGEST` followed by some sort of hash
- the second line, which is `F` followed by the absolute module name of the current file (e.g., `FCoq.Init.Datatypes`)
- Declarations of new definitions, which have the format
```
^([^ ]+) ([0-9]+):([0-9]+) ([^ ]+) ([^ ]+)$
```
where the first group is the type of the identifier being defined, the second and third groups define the character range, the fourth group is the module path within the file, and the fifth group is the unqualified name of the reference, or a way of describing the notation.

Disclaimer: all of this was reverse-engineered; none of it comes from me reading documentation.

* The full list, from `tools/coqdoc/index.ml`'s `type_of_string`, is:
```ocaml
let type_of_string = function
  | "def" | "coe" | "subclass" | "canonstruc" | "fix" | "cofix"
  | "ex" | "scheme" -> Definition
  | "prf" | "thm" -> Lemma
  | "ind" | "variant" | "coind" -> Inductive
  | "constr" -> Constructor
  | "indrec" | "rec" | "corec" -> Record
  | "proj" -> Projection
  | "class" -> Class
  | "meth" -> Method
  | "inst" -> Instance
  | "var" -> Variable
  | "defax" | "prfax" | "ax" -> Axiom
  | "syndef" -> Abbreviation
  | "not" -> Notation
  | "lib" -> Library
  | "mod" | "modtype" -> Module
  | "tac" -> TacticDefinition
  | "sec" -> Section
```

**Comment by ejgallego** [comment]
IMHO `.glob` files are not going to be very useful for this. SerAPI indeed can provide you many more information. Outputting a list of definitions using a given one and a map-like operation shouldn't be hard, just open a GH issue.
